src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineWrappingInBlockCheck.java

```diff
@@ -0,0 +1,440 @@
+ * the end of the block are configured separately.
+ * </p>
+ *
+ * <p> The check is for any statement that have a body:
+ * {@link TokenTypes#ANNOTATION_ARRAY_INIT ANNOTATION_ARRAY_INIT},
+ * {@link TokenTypes#ANNOTATION_DEF ANNOTATION_DEF},
+ * {@link TokenTypes#ARRAY_INIT ARRAY_INIT},
+ * {@link TokenTypes#CASE_GROUP CASE_GROUP},
+ * {@link TokenTypes#CLASS_DEF CLASS_DEF},
+ * {@link TokenTypes#CTOR_DEF CTOR_DEF},
+ * {@link TokenTypes#ENUM_CONSTANT_DEF ENUM_CONSTANT_DEF},
+ * {@link TokenTypes#ENUM_DEF ENUM_DEF},
+ * {@link TokenTypes#INSTANCE_INIT INSTANCE_INIT},
+ * {@link TokenTypes#INTERFACE_DEF INTERFACE_DEF},
+ * {@link TokenTypes#LAMBDA LAMBDA},
+ * {@link TokenTypes#LITERAL_CASE LITERAL_CASE},
+ * {@link TokenTypes#LITERAL_CATCH LITERAL_CATCH},
+ * {@link TokenTypes#LITERAL_DEFAULT LITERAL_DEFAULT},
+ * {@link TokenTypes#LITERAL_DO LITERAL_DO},
+ * {@link TokenTypes#LITERAL_ELSE LITERAL_ELSE},
+ * {@link TokenTypes#LITERAL_FINALLY LITERAL_FINALLY},
+ * {@link TokenTypes#LITERAL_FOR LITERAL_FOR},
+ * {@link TokenTypes#LITERAL_IF LITERAL_IF},
+ * {@link TokenTypes#LITERAL_NEW LITERAL_NEW},
+ * {@link TokenTypes#LITERAL_SWITCH LITERAL_SWITCH},
+ * {@link TokenTypes#LITERAL_SYNCHRONIZED LITERAL_SYNCHRONIZED},
+ * {@link TokenTypes#LITERAL_TRY LITERAL_TRY},
+ * {@link TokenTypes#LITERAL_WHILE LITERAL_WHILE},
+ * {@link TokenTypes#METHOD_DEF METHOD_DEF},
+ * {@link TokenTypes#OBJBLOCK OBJBLOCK},
+ * {@link TokenTypes#SLIST SLIST},
+ * {@link TokenTypes#STATIC_INIT STATIC_INIT}.
+ * </p>
+ *
+ * <p> By default the check will check the following statements:
+ *  {@link TokenTypes#ANNOTATION_DEF ANNOTATION_DEF},
+ *  {@link TokenTypes#CLASS_DEF CLASS_DEF},
+ *  {@link TokenTypes#ENUM_DEF ENUM_DEF},
+ *  {@link TokenTypes#INTERFACE_DEF INTERFACE_DEF}.
+ * </p>
+ *
+ * <p> An example of how to configure the check with default parameters is:
+ * </p>
+ *
+ * <pre>
+ * &lt;module name="EmptyLineWrappingInBlock"/&gt;
+ * </pre>
+ *
+ * <p> Example of declarations with empty line separator at the top of the block
+ * that is expected by the Check by default:
+ * </p>
+ *
+ * <pre>
+ * class Foo {
+ *
+ *     public int method() {
+ *
+ *         return 0;
+ *     }
+ * }
+ * </pre>
+ *
+ * <p> An example how to check wrapping of a {@link TokenTypes#CLASS_DEF CLASS_DEF} and
+ * {@link TokenTypes#INTERFACE_DEF INTERFACE_DEF} only:
+ * </p>
+ *
+ * <pre>
+ * &lt;module name="EmptyLineWrappingInBlock"&gt;
+ *    &lt;property name="tokens" value="CLASS_DEF, INTERFACE_DEF"/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ *
+ * <p>
+ * An example how to allow {@link BlockWrapOption#NO_EMPTY_LINE no empty line} after the left curly
+ * brace nor before the right curly brace:
+ * </p>
+ *
+ * <pre>
+ * &lt;module name="EmptyLineWrappingInBlock"&gt;
+ *    &lt;property name="topSeparator" value="no-empty-line"/&gt;
+ *    &lt;property name="bottomSeparator" value="no-empty-line"/&gt;
```

Comment:
EmptyLineWrappingInBlock vs topSeparator . Naming inconsistency.

Lets name check WrappingInBlock and properties "topWrapping", "bottomWrapping"

@rnveach , please review this proposal and confirm it or propose your names.

Reply:
maybe "topOption" and "bottomOption"?

Reply:
and "emptyBlockOption"

Reply:
> EmptyLineWrappingInBlock 
> WrappingInBlock 

LineBreakForBlock? I don't have a better name right now.
I don't like 'wrapping' as it is subjective what type of wrapping you are talking about, like instruction wrapping or the definition line's wrapping.

> topSeparator 
> topWrapping
> topOption

topBrace and bottomBrace or openingBrace and closingBrace.

Reply:
> LineBreakForBlock vs leftBrace and rightBrace

this is really good point, at it show misconception of original idea of Check - it tries to rule both concepts(for right and left curly) in one check.
Looks like we need two Checks - LineBreakForLeftCurly and LineBreakForRightCurly as we did for LeftCurly and RightCurly . So user will more consistent view.

@rnveach , @pbludov , does it make sense ?

Reply:
It makes sense but I don't really think it needs to be split into 2 right now. The check is pretty simple and this will also make configurations more complex as now instead of 1 definition for each type, we will need 2.

Reply:
Curly braces are divided because they have different policies. In this case, if we divide the check, this will only add confusion to the users.

Reply:
>LineBreakForLeftCurly and LineBreakForRightCurly

I do not like the idea of splitting the check into two almost identical. But I like those names.
Let it be "LineBreakAfterLeftCurly" and "LineBreakBeforeRightCurly".

Or may be "EmptyLineAfterLeftCurly" and "EmptyLineBeforeRightCurly" - to put it  much closer to the EmptyLineSeparator check in the documentation.

It that case the parameter must be the boolean. Simply "Need a blank" vs "Need no blanks". 
To have an enum for the boolean parameter is too overblown.

Please, give me a good name to this parameter.

Reply:
What about
```
<module name "AfterLeftCurly">
  <parameter name "needEmptyLine" value "true">
</module>
<module name "BeforeRightCurly">
  <parameter name "needEmptyLine" value "false">
</module>
```
?